Use Immutable `Record` for accounts in Redux state

renchap opened the pull request:
This is inspired by #25137, but uses proper `Record` objects everywhere, in addition to adding types.

Related to #26555 as well

I chose to introduce two new directories, to better organize the code:
- `api_types` contain the interfaces for the values returned by the Rails API. At some point, we will probably want to generate those using OpenAPI or equivalent
- `models` contain the type definitions for the objects used in the Redux state, as well as any associated functions. For now, they contain a function to create a model from an API JSON

Notable changes:
- `initialState` now uses the API JSON types when they are defined
- ~~`convertState` has been changed to create `Record` objects on keys they have corresponding models~~
- Converting a JSON from the API into a `Record` to be used in the state is all done in one place
- The `accounts_counters` reducer was merged into the `accounts` reducer
- `accounts` reducer and associated actions has been converted into Typescript
- `relationships` reduced has been converted into Typescript
- `makeGetAccount` is reimplemented to return a properly typed Record, allowing consumer components to be typed correctly

TODO:
- [ ] (maybe) add runtime checks (in dev only?) to ensure we do not have unknown fields in API responses that are dropped when creating the `Record`

renchap committed:
Convert `accounts` state to Immutable Record

renchap committed:
Improve `Account` models, move transformation from JSON into the model creator, convert the `accounts` reducer to Typescript

renchap commented:
Pushed the current state of my work.

- Object shapes should now be correct, both for the API JSONs and the Immutable `<Record>`, including computing the various fields that are used. It is all moved in the model
- `accounts` reducer is written in Typescript, and the actions are correctly typed. I could not use `createReducer` from RTK as it does not work well with an Immutable state, but it should not be hard to convert later
- `ACCOUNT_IMPORT` action was not used anywhere, dropping it
- `makeGetAccount` is reimplemented to return a properly typed `Record`. It is not yet complete as I need to convert `accountsCounter` and `relationships` to Typescript, but this should not be very hard

renchap committed:
Convert `<LimitedAccountHint>` to typescript

As it was the only component using `revealAction`

renchap committed:
Drop the `accounts_counters` reducer

Those are only used in `makeUserSelector()`, being merged back into `Account`. Easier to simply store them into `Account` directly.

renchap commented:
I dropped `accounts_counters` entirely in my latest commit, as the only use for this reducer is to be merged back into `Account` in `makeUserSelector`, let's store those directly in `Account`

renchap committed:
Fix the initial state type

trinitroglycerin commented:
One potential concern with mixing Redux Starter Kit and plain old reducers is that it's not easy to type it safely.

To type an "old-style" reducer, you need to create a discriminant union for `Action`:

```ts
type Action = 
  | { type: "ACTION_NAME"; payload: A[] }
  | { type: "ACTION_NAME_2"; payload: B[] }
```

With redux starter kit actions, you no longer have access to the string literal used to construct the action, which means you can't construct a discriminant without duplicating the name across files:

```ts
// app.ts
export const frobTheWidget = createAction('FROB_WIDGET');

// reducer.ts
type Action = 
  | { type: "FROB_WIDGET", ... }
  | { type: "ACTION_NAME"; payload: A[] }
  | { type: "ACTION_NAME_2"; payload: B[] }
```

Or by exporting it as a constant:

```ts
export const FROB_THE_WIDGET = "FROB_WIDGET";
export const frobTheWidget = createAction(FROB_THE_WIDGET);
```

You _can_ infer the type of the action struct:

```
type ActionType = ReturnType< typeof frobTheWidget>
```

However, the typings of RSK assigns the type `string` to the `type` property - which means that the discriminant union approach in a reducer no longer works, since _any_ string is assignable to `string`:

```
function reduce(state, action) {
   switch (action.type) {
      case frobTheWidget.type:
        // Because frobTheWidget is assignable to string it has access to `payload` which can be `A[]` or `B[]`!
      case "ACTION_NAME":
        // Because ACTION_NAME is assignable to string, `action` is narrowed to `{ payload: A[] } | {} ` - Therefore, `payload` may not exist 
   }
}
```

For the purposes of this PR, this presents in `accounts_map.js`, `relationships.js` and `timelines.js` as of 5352ca1; these reducers cannot be safely typed if they mix RSK actions and POJO actions.

renchap commented:
If the reducers are not converted to Typescript, you can use `frobTheWidget.type` like in your example.

Once they are converted to Typescript, you can use either:
- `createReducer` from RTK, if the state has been converted to plain object
- `else if(fromTheWidget.match(action)) { … }` if the state is still an Immutable object, like I did in this PR with `reducers/account.ts`

In both cases, all the actions needs to be converted to `createAction`, but I do not think this is a big issue?

trinitroglycerin commented:
Yes, sorry, my comments are _not_ intended to be a criticism. When merging this code into the branch for the PR I noticed the typing issue that might occur, so I wanted to mention it.

renchap commented:
No problem, I was only wondering if you saw an issue with the approach I described 😄 

I plan to have a go at converting the `relationships` reducer to Typescript, and all the actions it uses to `createAction`, before marking this PR as ready.

Also this will probably need to wait for the 4.2 release branch cutoff (hopefully next week) before being merged, we do not really want to merge such a big change just before a release.

trinitroglycerin commented:
Aligned! I'll plug away on some other component refactors :)

renchap committed:
Convert `account_notes` actions to Typescript

renchap committed:
Create Relationship model & associated types

renchap committed:
Convert actions used in `relationshipReducer` to Typescript

renchap committed:
Convert `relationships` reducer to Typescript

I needed to revert to `createAppAsyncThunk` due to a circular Redux type problem, I need to check how to fix it properly

renchap commented:
Relationships should now be fully typed.

I also split a few of the commits from this PR into their own PRs, to make it easier to review.

There is a type issue with using `createAppAsyncThunk` once a reducer starts referencing it, as it creates a circular dependency issue.

I need to check why it does this.

renchap committed:
Force the `Reducer` type to avoid circular type dependencies and allow usage of `createAppAsyncAction`

renchap committed:
Small fixes

renchap commented:
This should be ready for review :)

renchap committed:
Fix linting

ClearlyClaire commented on app/javascript/mastodon/models/account.ts:
```diff
@@ -0,0 +1,149 @@
+      serverJSON.fields.map((field) => createAccountField(field, emojiMap)),
+    ),
+    emojis: List(serverJSON.emojis.map((emoji) => CustomEmojiFactory(emoji))),
+    roles: List(serverJSON.roles.map((role) => AccountRoleFactory(role))),
```
`roles` is optional (`app/javascript/mastodon/api_types/accounts.ts` seems wrong about it), making this break when dealing with remote accounts

renchap committed:
Mark more fields as optional

renchap committed:
Remove unneeded change

ClearlyClaire commented on app/javascript/mastodon/actions/accounts_typed.ts:
```diff
@@ -0,0 +1,97 @@
+  id: string;
+}>('accounts/revealAccount');
+
+export const importAccounts = createAction<{ accounts: ApiAccountJSON[] }>(
+  'accounts/importAccounts',
+);
```
I'm noticing that `importAccounts` used to work on normalized (internal) objects but was changed to do the normalization later in the process (as part of the reducer). What is the rationale behind this change?
